### PR TITLE
chore: added checkin_cli test to federated integ test

### PR DESCRIPTION
**Pull request by uri200**
Signed-off-by: Oriol Batalla <obatalla@fb.com>

## Summary

Enhanced federated integ test to add checkin cli command for connectivity test

## Test Plan

fab test_all

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/integ_tests/federated_tests/README.md [hunk 1]
```diff
@@ -6,87 +6,93 @@ hide_title: true
...
+
+| Services          |  Vagrant VM       |  Docker   |
+|-------------------|:-----------------:|:----------:|
+| AGW               | magma             |           |
+| FeG               |                   | &check;   |
+| Orc8r             |                   | &check;   |
+| Traffic server    | magma_trfserver   |           |
+| S1AP tester       | magma_test        |           |
+
+
```
Comment (uri200) [inline lte/gateway/python/integ_tests/federated_tests/README.md:30]: image

File: lte/gateway/dev_tools.py [hunk 2]
```diff
@@ -137,6 +139,28 @@ def deregister_feg_gw():
...
 
+def check_agw_cloud_connectivity(timeout=10):
+    """
+    Check connectivity of AGW with the cloud using checkin_cli.py
+    Args:
+        timeout: amount of time the command will retry
+    """
+    vagrant_setup("magma", False, force_provision=False)
+    with cd("/home/vagrant/build/python/bin/"):
+        dev_utils.run_fab_command_with_repetition("./checkin_cli.py", timeout)
```
Comment (hcgatewood) [inline lte/gateway/dev_tools.py:178]: Nit here and below: can coalesce these into single line, by specifying the executable's full path

File: lte/gateway/python/integ_tests/federated_tests/README.md [hunk 3]
```diff
@@ -6,87 +6,93 @@ hide_title: true
...
+and uses `S1aP` and `magma_trfserver` as an eNb and SGi emulator to run tests.
 
 ```mermaid
 graph LR;
-    S1AP --> AGW --> orc8r --> FEG --> 3ggp_CORE;
-    3ggp_CORE --> FEG --> orc8r --> AGW --> S1AP;
-    AGW --> trf_server;
-    trf_server --> AGW;
+    A[[S1AP]] --- AGW ---|control| orc8r --- FEG --- HSS[(HSS)];
+    AGW ---|data| Z[[magma_trfserver]];
```
Comment (hcgatewood) [inline lte/gateway/python/integ_tests/federated_tests/README.md:17]: Wow that's so pretty / looks so easy to write. Do you have a pointer to the doc/tutorial you used to understand mermaid?

File: lte/gateway/python/integ_tests/federated_tests/README.md [hunk 4]
```diff
@@ -6,87 +6,93 @@ hide_title: true
...
+| FeG               |                   | &check;   |
+| Orc8r             |                   | &check;   |
+| Traffic server    | magma_trfserver   |           |
+| S1AP tester       | magma_test        |           |
+
+
+## Build environment
+
+Next steps show you how to build the test environment automatically. If you
+want to build this step by step go to [Manual build](#Manual build)
```
Comment (hcgatewood) [inline lte/gateway/python/integ_tests/federated_tests/README.md]: Does `#Manual build` work correctly? I think it might have to be `#manual_build` but not sure

File: orc8r/tools/fab/dev_utils.py [hunk 5]
```diff
@@ -405,3 +407,43 @@ def cloud_delete(
...
+    """
+    run command on remote machine using fabric.api.run. Repeats on error
+    Args:
+        command: command to run
+        timeout: time to execute the command while it fails
+    """
+    _command_with_repetition(run, command, timeout)
+
+
+def _command_with_repetition(func, command, timeout=5):
```
Comment (hcgatewood) [inline orc8r/tools/fab/dev_utils.py:444]: Nit: "timeout" gives the connotation it's time-based. Maybe something like "max retries"?

**Review by hcgatewood — APPROVED** [review]
Awesome! Couple small notes but lgtm

File: lte/gateway/python/integ_tests/federated_tests/README.md (hunk 3 above)
Comment (uri200) [reply lte/gateway/python/integ_tests/federated_tests/README.md:17]: https://mermaid-js.github.io/mermaid/#/./flowchart?id=flowcharts-basic-syntax

File: lte/gateway/dev_tools.py (hunk 2 above)
Comment (uri200) [reply lte/gateway/dev_tools.py:178]: Since fab it is a command line automation format, it is better to have each step separated so it is more readable and easier to debug.

It also provides more granularity in case of error (if that folder doesn't exist you will get an explicit error)

File: orc8r/tools/fab/dev_utils.py (hunk 5 above)
Comment (uri200) [reply orc8r/tools/fab/dev_utils.py:444]: In fact it is a time out.  We are not worried for the number of attempts (which happen every second)

The reason for that is on those integ test normally the system is not ready by the time the test starts. Adding this allow us to have some control on how long will the test retry.  This command does not pretend to be exact (if it takes more tjhan 5 seconds is fine, it will not take less in case the system is down)

In other words, this just mimics what a person would do while waiting for the service to start (similar to running a watch command)

If i am able to add this tests on github actions , I will call that command with 60 to make sure I give the system plenty of time to start and bootstrap

Comment (hcgatewood) [reply orc8r/tools/fab/dev_utils.py:444]: Hm if I'm reading correctly, it looks like there's no timing functionality specifically in this function -- looks like it increments a count after every attempt, and after a max number of retires gives up

To me, timeout means "try for at most x seconds", while max tries means "try at most this number of times". Not a huge deal, and maybe I'm reading it wrong 🙂

Comment (uri200) [reply orc8r/tools/fab/dev_utils.py:444]: check line 459

Comment (hcgatewood) [reply orc8r/tools/fab/dev_utils.py:444]: Right right. I guess the consideration is still: consider if "func" takes e.g. 10 minutes to run -- then the total runtime would be `timeout*10m1s`, rather than `timeout*1s`. Which makes it feel like a max_retries rather than timeout

But again, just a nit, up to you!

**Review by themarwhal — APPROVED** [review]
some of the pep8 annotations seem like they are worth addressing but other than that lgtm!

File: orc8r/tools/fab/dev_utils.py (hunk 5 above)
Comment (uri200) [reply orc8r/tools/fab/dev_utils.py:444]: Yes I considered that, but since the commands used by the fab command here are normally limited in time I didn't see it necessary. But since the code is in a shared library it may be wise to add an upper bound. 

Will change some things so we can make the function more generic and make sure we don't exceed the time out

**Review by koolzz — APPROVED** [review]

File: lte/gateway/python/integ_tests/federated_tests/README.md [hunk 6]
```diff
@@ -6,87 +6,93 @@ hide_title: true
...
-
-Once it is built, start the traffic and test vm
+## Run test
+Once you have built all the VMs, you can try to run a test from
+`magma_test` VM
 ```bash
 cd magma/lte/gateway
-vagrant up magma_test
-vagrant up magma_trfserver
+vagrannt ssh magma_test
```
Comment (ardzoht) [inline lte/gateway/python/integ_tests/federated_tests/README.md]:
```suggestion
vagrant ssh magma_test
```

**Review by ardzoht — APPROVED** [review]
minor nit on docs typo

File: .github/workflows/federated-integ-test.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
+---
+
+name: Federated integ test
+
+on:  # yamllint disable-line rule:truthy
+  push:
+    branches:
+      - master
+      - v1.*
+  pull_request:
```
Comment (quentinDERORY) [inline .github/workflows/federated-integ-test.yml]: we don't want heavy integration tests to be triggered on pull_request.
You should use the workflow_dispatch trigger to execute on demand on PR.

**Review by quentinDERORY — CHANGES_REQUESTED** [review]
LGTM but need to change trigger on GHA job

File: .github/workflows/federated-integ-test.yml (hunk 7 above)
Comment (uri200) [reply .github/workflows/federated-integ-test.yml]: yep, this is a test right now. I will probably remove the whole file once i get this running (if i get it)

Will create a separate pr for that

File: .github/workflows/federated-integ-test.yml [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        id: date
+      - uses: actions/checkout@v2
+      - name: setup pyenv
+        uses: "gabrielfalcao/pyenv-action@v8"
+        with:
+          default: 3.8.5
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.8.5'
+      - name: Install Docker
```
Comment (quentinDERORY) [inline .github/workflows/federated-integ-test.yml]: Use this module to install docker and docker-compose : https://github.com/marketplace/actions/setup-docker

File: .github/workflows/federated-integ-test.yml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+          docker save mariadb:latest  | gzip > mariadb.tar.gz
+          docker save feg_gateway_go:latest  | gzip > feg_gateway_go.tar.gz
+          docker save feg_gateway_python:latest  | gzip > feg_gateway_python.tar.gz
+      - uses: actions/upload-artifact@v2
+        with:
+          name: federated-integ-test-docker-images
+          path: images
+
+  federated-integ-test:
+    runs-on: macos-10.15
```
Comment (quentinDERORY) [inline .github/workflows/federated-integ-test.yml]: add 
`needs: docker-build-federated-integ-test`

File: .github/workflows/federated-integ-test.yml (hunk 7 above)
Comment (uri200) [reply .github/workflows/federated-integ-test.yml]: Hey @quentinDERORY  I removed the github action file form this PR

File: lte/gateway/python/integ_tests/federated_tests/docker/docker-compose.override.yml [hunk 10]
```diff
@@ -1,27 +1,6 @@
...
-      - gwcerts:/var/opt/magma/certs
-      - gwconfigs:/var/opt/magma/configs
-    command:
-      - /bin/bash
-      - -c
-      - |
-         mkdir -p ../../.cache/feg/
-         touch ../../.cache/feg/snowflake
-         tail -f /dev/null
-
```
Comment (uri200) [inline lte/gateway/python/integ_tests/federated_tests/docker/docker-compose.override.yml:24]: test service is not needed for federated integ test

**Review by quentinDERORY — APPROVED** [review]
